### ISSUE TITLE
Support latex labels

### ISSUE THREAD

**Issue by dawehner** [bokeh/bokeh#647]
It would be great if bokeh would support latex labels either rendered via a native latex installation of mathjax, if available. As you probably all know matplotlib has a functionality like that which makes it really conventient for the scientitic community: http://matplotlib.org/1.3.1/users/usetex.html

Possible syntax:

````
import bokeh.plotting as bplt
import numpy as np
bplt.output_notebook()

bplt.scatter(np.random.randn(100), np.random.randn(100))
bplt.xaxis().axis_label=r'$f(x)$'
bplt.show()
````

**Comment by gully** [comment]
+1 on this feature request.

**Comment by chuyelchulo** [comment]
I'll add a +1 here as well.

**Comment by mcaffini** [comment]
:+1:

**Comment by cjtonde** [comment]
+1 for this feature request.

**Comment by MaxNoe** [comment]
+1

**Comment by pycckuu** [comment]
+1

**Comment by bryevdv** [comment]
ping @bokeh/core just when I was about to give up on native direct-to-canvas rendering for LaTeX, someone dangles out a hope: https://github.com/Khan/KaTeX/pull/246#issuecomment-112947054

**Comment by gagern** [comment]
Feel free to follow development on [this branch](https://github.com/gagern/KaTeX/commits/canvas). The idea is to take KaTeX's internal DOM representation (which isn't HTML DOM yet, but translates to that literally), and then reproduce everything the style (from CSS classes as well as local attributes) does to it, emitting canvas rendering instructions instead. We have to make that two-phase, since KaTeX works without width information, so we have to measure boxes and then adjust horizontal positioning accordingly. Getting all the fonts loaded early on seems a bit tricky. I'll be happy for anyone who double-checks existing code or ports more layouting details from CSS to JavaScript. If you want to contribute, feel free to post pull requests into my branch.

Since @bryevdv asked for an estimated time frame: I expect to spend about a week on this, among other things. Everything that works after the week is good. What does not work is either too exotic for me to care, or too difficult to get implemented, or whatever. If other things keep me busy, it might be two weeks before I try to get this merged into KaTeX, i.e. before I start formulating test cases, screenshot procedures and so on, and then post a pull request.

**Comment by fpliger** [comment]
thanks @gagern , this is super exciting!

**Comment by damianavila** [comment]
Yep... really great, thanks @gagern for your work on this...

**Comment by profxj** [comment]
+1

**Comment by karenyyng** [comment]
+1

**Comment by benlaken** [comment]
+1

**Comment by altosaar** [comment]
+1!

**Comment by smsaladi** [comment]
+1

**Comment by pablocobelli** [comment]
+1

**Comment by bryevdv** [comment]
Just an update:, there is a preliminary PR to add a proper label annotation here: https://github.com/bokeh/bokeh/pull/3956 Once this is complete (for basic text) we can look to extending it using KaTeX. Please note that since KaTeX will render to DOM elements (divs) on top of the canvas, it is very likely that math text and static output (png, svg) will *not be compatible* with each other. I'm just not sure what can be done about that, unfortunately. If there are ideas to make that work, they are certainly appreciated.

**Comment by fedhere** [comment]
+1

**Comment by jacopo-chevallard** [comment]
+1

**Comment by ianhbell** [comment]
+1

**Comment by yarps** [comment]
+1 really needed for my project

**Comment by gagern** [comment]
Have you tried using [my render to canvas branch](https://github.com/Khan/KaTeX/pull/251) for this? If it works, we might try convincing KaTeX devs to merge something like that. If not, I'd like to know where it fails. I'm [using that approach](https://github.com/CindyJS/CindyJS/blob/c3cca35ecd936cadb81ef6b1610de12e34f303b1/plugins/katex/src/js/katex-plugin.js#L151) in CindyJS without major problems so far.

**Comment by birdsarah** [comment]
Some progress here: https://github.com/bokeh/bokeh/pull/4914

**Comment by canavandl** [comment]
@gagern  I'll make a note to experiment with your branch over the weekend. 

Our primary issue was deciding how to include arbitrary external JS/CSS resources in our output. But we think we've figured that out in #4914 and made a small example using katex.renderToString and superpositioning the resulting span.

**Comment by bryevdv** [comment]
This was added via an example of a custom extension in `0.12.2` any futher enhancements should get their own specific issues

**Comment by MaxNoe** [comment]
Maybe it's just me, but I think this should be a native feature, not a custom extension.

**Comment by birdsarah** [comment]
@MaxNoe, we're working hard to trim 19kb here, 20kb there from BokehJS to make the core library something that's as lightweight as possible so folks don't have to pay for code they don't need.

Putting this into core would add, if I understand correctly, at least 30kb in minified code. From my point of view, this is too specialized a need to justify this increase in size to the library.

Having said that, I think this would be a great candidate for an "official plugin". "Official plugin" isn't a real thing yet, but we're planning something for the geo features that are currently in the BokehJS library - but shouldn't be for the same reasons. There is a write-up of the plan here: https://github.com/bokeh/bokeh/issues/4715. 

I would suggest opening a new issue to discuss this further, perhaps linking back to here, as all the plus ones on this issue well document the community interest in the feature.

I hope you understand this fine balance and our plan to strike a balance between making commonly requested, yet still specialized, features not part of core, but still easy to install and use.

**Comment by MaxNoe** [comment]
I'm a  physicist, for me, the proper rendering of math is at the very  core of a plotting library.

**Comment by birdsarah** [comment]
I understand that completely. What I'm trying to say is that physicists are not our only users and we want to meet the diverse needs of our community as best as possible.

**Comment by MaxNoe** [comment]
Then I really would like to see a plugin system, which makes that easy. The quoted example is not something which makes me think: "Oh this is solved now". I would strongly opt for this issue to be reopened.

**Comment by bryevdv** [comment]
@MaxNoe A custom extension is as-solved as this can reasonably be, given the different requirements of a large and varied user base. Making extensions easily discoverable, installable, and sharable among the community is already covered by another issue as @birdsarah mentions. 

Now if there are other specific refinements or things to demonstrate, that's fine. However, those requests should go in *new* and *specific* issues. "Add LaTeX" support is too broad, means different things to different people. This issue can at most be about getting a foothold of capability to build on. Further feature request issues need to be as narrow and specific as possible.

**Comment by MaxNoe** [comment]
Are we talking about this example: http://bokeh.pydata.org/en/dev/docs/user_guide/extensions_gallery/latex.html?

This is nowhere near the first asked question in this issue: rendering LaTeX in axis labels.
> bplt.xaxis().axis_label=r'$f(x)$'

**Comment by bryevdv** [comment]
@MaxNoe Another thing worth mentioning, the custom extension facility *is* the plugin system. You can  take the code in the example, and drop it into your project as-is and use that `LatexLabel` immediately. Is there is still work to do to make extensions more easily sharable? Absolutely, but my own personal view is that it is far better to expose and advertise the capability sooner, rather than wait even longer. 

Regarding this issue, it dates from an early time in the library when "big grand (and vague)" features helped drive fast and furious development by a couple of individual contributors. Bokeh is at a very different stage now. We have to be very concerned with long term stability and maintainability. To that end we have to proceed from *narrow*, *focused* issues that have *concrete proposals* and *clear end criteria*. I hope you will agree this issue is not that. I encourage to you make other more specific suggestions to build upon the capability that has been demonstrated in `0.12.2`

Bokeh is already an *enormous* library with a huge testing surface. We need to find ways to make it smaller and more stable. Offering the extension mechanism is exactly for that purpose, to allow people to do all the many things they might reasonably want to do, that might never make sense to put in the core library, or that the extremely limited team resources simply won't realistically have time time (we do all have friends and families and other jobs, after all).

**Comment by joshburkart** [comment]
@bryevdv Thanks for all your hard work on this great library! Do you agree, though, that it should eventually be as simple as a couple lines to get a LaTeX label for a given plot? How do you suggest tracking/planning development of this rather essential feature if not via a GitHub issue?

Also, isn't this required to get e.g. log plots to show e.g. the 2 in 10^2 properly, rather than just printing literally "10^2"? I'm looking [here](http://bokeh.pydata.org/en/latest/docs/user_guide/plotting.html#log-scale-axes).

**Comment by bryevdv** [comment]
I'm not saying things should not be tracked in GitHub issues. I am saying *focused*, *specific*,  *concrete* feature proposals with *clear success criteria* should be tracked in *separate*, *individual* issues, so that the *extremely limited* team bandwidth can be *managed and organized effectively and efficiently*, and that this particular issue in its current form is not useful for the project in that regard (therefore it should be closed).

Getting exponents in log plots to show up correctly is going to take quite alot of heavy lifting and refactoring of `Axis` first, to support the ability to use true `Label` subclasses rather than just draw everything naively itself. All of that is completely incidental to any sort of LaTeX integration, but that's exactly the point. These things need to be broken down into manageable, trackable, digestible chunks. "Add LaTeX support" is *not that*.  "Add LaTeX support" has served its purpose. What we have now is a basis and a foundation. Let's build upon that in a realistic and incremental fashion with actionable and measurable issues. 

Finally, I do not think that 
```
bplt.xaxis().axis_label=r'$f(x)$'
```
alone, by itself, is ever going to be realistic. We absolutely *do no have* the luxury that a "render once in python straight to an image" library has, by definition. With Bokeh, Python is not where *any* of the *actual work* happens. All the real work happens in the browser. That is completely unlike, say, MPL. This difference provides a unique set of new capabilities, but also has it's own set of challenges and drawbacks. There are *always tradeoffs*. 
```
from something import LatexLabel, LatexTickFormatter  # maybe something = bokeh.ext 

p.xaxis.axis_label = LatexLabel('$f(x)$')
p.xaxis[0].formatter = LatexTickFormatter()
```
is my personal opinion for what a reasonable target could be.

**Comment by arossi1** [comment]
I noticed in the 0.12.3 CHANGELOG that this feature was added:
#647 Support latex labels

I tried using LaTeX on the axis label, but it does not render.  Does anyone know if there is special syntax to enable support?  I don't see anything in the docs.  Thanks in advance!

**Comment by bryevdv** [comment]
There is more information in the blog post: https://bokeh.github.io/blog/2016/9/6/release-0-12-2/

**Comment by MaxNoe** [comment]
@arossi1 No, this feature was not added. An example was added how you can write a bunch of custom code it implement it yourself.

**Comment by arossi1** [comment]
Thanks for the info.  I see the example:
http://bokeh.pydata.org/en/latest/docs/user_guide/extensions_gallery/latex.html#userguide-extensions-examples-latex

...but I get this error:

Bokeh Error
Module `LatexLabel' does not exists. The problem may be two fold. Either a model was requested that's available in an extra bundle, e.g. a widget, or a custom model was requested, but it wasn't registered before first usage.

Do you know of a full example, I'm probably doing something incorrectly.

Thanks!

**Comment by bryevdv** [comment]
@arossi1 That is a complete example. I just ran it with Bokeh `0.12.3` and it functioned as expected. Please note that custom models do no currently work (well) with `bokeh.client` and `push_session` if that is your case. There are workarounds, but the best path path forward is still being discussed. If this is not your case, please open a new, separate  bug issue. 

@MaxNoe An example was added that you can use practically verbatim. That is: all the "custom code" was *written for you*. 

If Bokeh does not suit your current needs, please feel free to:

* pitch in and help make it better yourself [1]
* use something else that help you with your particular problem. 
 
[1] of course code is always welcome, but see above even regarding focused concrete proposals, which no one has yet to create.

**Comment by arossi1** [comment]
Sorry, I deleted my previous comment because I realized the example works stand-alone, but not when I put it into my project, which is rendering a script/div via the "bokeh.embed.components" API.  I searched around for some examples, and came across a custom tickmark example and noticed they are adding the formatter:

`p.xaxis[0].formatter = DateGapTickFormatter()`

Do I need to do something similar in the "latex label" case?

**Comment by bryevdv** [comment]
ping @mattpap

**Comment by arossi1** [comment]
Thanks @bryevdv  -- I've been looking around some more and this has been informative:
https://github.com/bokeh/bokeh/pull/4914

In general, I don't see the LatexLabel class included in the resulting html javascript section.  I tried manually injecting it and I no longer get the error:

_Bokeh Error
Module `LatexLabel' does not exists. The problem may be two fold. Either a model was requested that's available in an extra bundle, e.g. a widget, or a custom model was requested, but it wasn't registered before first usage._

However, my plot does not render.

It seems like I need to make a connection between the LatexLabel class defined in python (with __javascript__, __css__, and __implementation__ attributes) and the Bokeh figure / plot.  I'll keep searching for a place to wire it up.

Thanks for your help!

**Comment by bryevdv** [comment]
@arossi1 extension code should be included automatically. Can you file a separate GH bug report with all relevant information?

**Comment by arossi1** [comment]
Sure, I'll put a minimal example together and submit a GH bug report.  Thanks for your help.

**Comment by arossi1** [comment]
Sorry for the delay, I added issue https://github.com/bokeh/bokeh/issues/5353
